src/annotator/integrations/vitalsource.js

```diff
@@ -0,0 +1,208 @@
+ *
+ * This integration delegates to the standard HTML integration for most
+ * functionality, but it adds logic to:
+ *
+ *  - Customize the document URI and metadata that is associated with annotations
+ *  - Prevent VitalSource's built-in selection menu from getting in the way
+ *    of the adder.
+ *
+ * @implements {Integration}
+ */
+export class VitalSourceContentIntegration {
+  /**
+   * @param {HTMLElement} container
+   */
+  constructor(container = document.body) {
+    this._htmlIntegration = new HTMLIntegration(container);
+
+    this._listeners = new ListenerCollection();
+
+    // Prevent mouse events from reaching the window. This prevents VitalSource
+    // from showing its native selection menu, which obscures the client's
+    // annotation toolbar.
+    //
+    // VitalSource only checks the selection on the `mouseup` and `mouseout` events,
+    // but we also need to stop `mousedown` to prevent the client's `SelectionObserver`
+    // from thinking that the mouse is held down when a selection change occurs.
+    // This has the unwanted side effect of allowing the adder to appear while
+    // dragging the mouse.
+    const stopEvents = ['mousedown', 'mouseup', 'mouseout'];
+    for (let event of stopEvents) {
+      this._listeners.add(document.documentElement, event, e => {
+        e.stopPropagation();
+      });
+    }
+  }
+
+  canAnnotate() {
+    return true;
+  }
+
+  destroy() {
+    this._listeners.removeAll();
+    this._htmlIntegration.destroy();
+  }
+
+  /**
+   * @param {HTMLElement} root
+   * @param {Selector[]} selectors
+   */
+  anchor(root, selectors) {
+    return this._htmlIntegration.anchor(root, selectors);
+  }
+
+  /**
+   * @param {HTMLElement} root
+   * @param {Range} range
+   */
+  describe(root, range) {
+    return this._htmlIntegration.describe(root, range);
+  }
+
+  contentContainer() {
+    return this._htmlIntegration.contentContainer();
+  }
+
+  fitSideBySide() {
+    // Not yet implemented
+    return false;
+  }
+
+  async getMetadata() {
+    // Return minimal metadata which includes only the information we really
+    // want to include.
+    return {
+      title: document.title,
+      link: [],
+    };
+  }
+
+  async uri() {
+    // An example of a typical URL for the chapter content in the Bookshelf reader is:
+    //
+    // https://jigsaw.vitalsource.com/books/9781848317703/epub/OPS/xhtml/chapter_001.html#cfi=/6/10%5B;vnd.vst.idref=chap001%5D!/4
+    //
+    // Where "9781848317703" is the VitalSource book ID ("vbid"), "chapter_001.html"
+    // is the location of the HTML page for the current chapter within the book
+    // and the `#cfi` fragment identifies the scroll location.
+    //
+    // Note that this URL is typically different than what is displayed in the
+    // iframe's `src` attribute.
+
+    // Strip off search parameters and fragments.
+    const uri = new URL(document.location.href);
+    uri.search = '';
+    return uri.toString();
+  }
```

Comment:
Do you anticipate adding any more nuance here, or is removing `queryString` dependable for creating the right VS URI?

Reply:
I'm pretty sure removing the whole query string will be fine, since the path clearly identifies the content.